python/dolfinx/nls/petsc.py

```diff
@@ -72,3 +91,453 @@ def setP(self, P: types.FunctionType, Pmat: PETSc.Mat):  # type: ignore
+        matrix_creator = create_matrix
+        vector_creator = create_vector
+    elif assembly_type == fem.AssemblyType.block:
+        matrix_creator = create_matrix_block
+        vector_creator = create_vector_block
+    elif assembly_type == fem.AssemblyType.nest:
+        matrix_creator = create_matrix_nest
+        vector_creator = create_vector_nest
+    else:
+        raise ValueError("Unsupported SNES type")
+    A = matrix_creator(a)  # type: ignore
+    b = vector_creator(L)  # type: ignore
+    x = vector_creator(L)  # type: ignore
+    P = None if P is None else matrix_creator(P)  # type: ignore
+    return A, x, b, P
+
+
+class SNESSolver:
+    def __init__(
+        self,
+        F: typing.Union[dolfinx.fem.Form, ufl.form.Form],
+        u: dolfinx.fem.Function,
+        J: typing.Optional[typing.Union[dolfinx.fem.Form, ufl.form.Form]] = None,
+        P: typing.Optional[typing.Union[dolfinx.fem.Form, ufl.form.Form]] = None,
+        assembly_type: fem.AssemblyType = fem.AssemblyType.default,
+        bcs: typing.Optional[list[dolfinx.fem.DirichletBC]] = None,
+        form_compiler_options: typing.Optional[dict] = None,
+        jit_options: typing.Optional[dict] = None,
+    ):
+        """Initialize a PETSc-SNES solver."""
+        self._u = u
+        self._snes, self._x = create_snes_solver(
+            F, self._u, J, P, assembly_type, bcs, form_compiler_options, jit_options
+        )
+        if assembly_type == fem.AssemblyType.default:
+            self._copy_solution = copy_solution
+            self._replace_solution = replace_solution
+        elif assembly_type == fem.AssemblyType.block:
+            self._copy_solution = copy_solution_block
+            self._replace_solution = replace_solution_block
+        elif assembly_type == fem.AssemblyType.nest:
+            self._copy_solution = copy_solution_nest
+            self._replace_solution = replace_solution_nest
+        else:
+            raise ValueError("Unsupported Assembly type")
+
+    def set_options(self, options: dict):
+        opts = PETSc.Options()  # type: ignore
+        for key, v in options.items():
+            opts[key] = v
+        self._snes.setFromOptions()
+        # Delete options from handler post setting
+        for key, _ in options.items():
+            del opts[key]
+
+    def solve(self) -> tuple[PETSc.Vec, int, int]:  # type: ignore
+        """Solve the problem and update the solution in the problem instance.
+
+        Returns:
+            The solution, convergence reason and number of iterations
+        """
+
+        # Move current iterate into the work array.
+        self.copy_solution(self._u, self._x)
+
+        # Solve problem
+        self._snes.solve(None, self._x)
+
+        converged_reason = self._snes.getConvergedReason()
+        self.replace_solution(self._u, self._x)
+        return self._x, converged_reason, self._snes.getIterationNumber()
+
+    def __del__(self):
```

Comment:
This class design is now very nice, but the issue to decide on is whether we want to smoothen out the solve and memory management in the first place. Another option would be to provide a high-level NonLinearProblem class, in the same style as LinearProblem.

Reply:
I agree, so in effect I'm saying there is an argument this should just be called `NonlinearSolver` which wraps low-level SNES routines, in the same way `LinearSolver` wraps low-level KSP routines.

Reply:
I think we could remove the interface to `NewtonSolver` from Python, as I don't really see the benefit of it any longer. It is just a very naive implementation of `Newton` that somewhat interfaces with PETSc.

Reply:
It is a side-track from the PR at hand. Whatever we want to do with NewtonSolver (make it purely native for instance) would require access to mumps or superlu dist. Which we currently do not have.

The current NewtonSolver is tied to PETSc, and is used in C++ by people not wanting to set up their own PETSc SNES solver there, so lets keep it for a rainy day.

Reply:
This is now resolved in favour of offering a NonlinearProblem class aimed at beginner users, in the same style as LinearProblem.